File: cpp/Heap.cpp
#include <iostream>
#include <vector>
#include <stdio.h>
#include <math.h>
#include <algorithm>

using namespace std;


class Heap {
	protected:
		vector<int> heap;
	public:
		Heap() {
			heap = vector<int>();
		}

		static int get_left_child_index(int n) {
			return 2*n + 1;
		}

		static int get_right_child_index(int n) {
			return 2*n + 2;
		}

		static int get_parent_index(int n) {
			return floor((n - 1) / 2);
		}

		void print_heap() {
			int n = heap.size();

			for (int i=0; i<n; i++) {
				cout << heap[i] << " ";
			}

			cout << endl;
		}

		int get_size() {
			return heap.size();
		}
};


class MinHeap : public Heap {
	public:
		MinHeap() {
			Heap();
		}


		/**
		* 	Push the element to the last spot in the heap, then iteratively swap with parent elements if parents are smaller
		*/
		void push(int value) {
			heap.push_back(value);

			int index = heap.size() - 1;
			int parent_index = Heap::get_parent_index(index);

			while (heap[index] < heap[parent_index] && parent_index >= 0) {
				swap(heap[index], heap[parent_index]);
				index = parent_index;
				parent_index = Heap::get_parent_index(index);
			}
		}


		/**
		*	To pop, remove top element and replace with last
		*	Bubble down. Basically, compare element to both children.
		*	If no children, stop
		*	If one child and child smaller, swap
		*	If two children, if either are smaller, swap with smallest. If neither, don't swap at all
		*/
		int pop() {
			int top = heap.front();
			int temp = heap.back();
			heap.pop_back();
			heap[0] = temp;

			int index = 0;
			int left_child_index = Heap::get_left_child_index(index);
			int right_child_index = Heap::get_right_child_index(index);

			while (left_child_index < heap.size()) {

				// If only the node only has a left child, check if child is smaller than parent. If so, swap
				if (right_child_index >= heap.size()) {
					if (heap[left_child_index] < heap[index]) {
						swap(heap[left_child_index], heap[index]);
						index = left_child_index;
					} else {
						break;
					}
				}
				// Otherwise, check which of the children is smaller and swap with that child (if that child is smaller than the parent)
				else {
					if (heap[left_child_index] < heap[right_child_index] && heap[left_child_index] < heap[index]) {
						swap(heap[left_child_index], heap[index]);
						index = left_child_index;
					} else if (heap[right_child_index] < heap[left_child_index] && heap[right_child_index] < heap[index]) {
						swap(heap[right_child_index], heap[index]);
						index = right_child_index;
					} else {
						break;
					}
				}

				left_child_index = Heap::get_left_child_index(index);
				right_child_index = Heap::get_right_child_index(index); 
			}

			return top;
		}
};


int main() {
	MinHeap heap = MinHeap();
	heap.push(1);
	heap.push(0);
	heap.push(100);
	heap.push(20);
	heap.push(99);
	heap.push(-11);
	heap.push(22);
	heap.push(10);

	heap.print_heap();

	// This is heap sort: pop elements from a heap one by one
	int size = heap.get_size();
	for (int i = 0; i < size; i++) {
		cout << heap.pop() << " ";
	}
	cout << endl;

	return 0;
}

File: python/Heap.py
# Heap
from collections import deque
import math

class MinHeap(object):
	heap = None

	def __init__(self):
		self.heap = deque()

	def getParentIndex(self, ind):
		return math.floor((ind - 1) / 2)

	def getLeftChildIndex(self, ind):
		return 2 * ind + 1

	def getRightChildIndex(self, ind):
		return 2 * ind + 2

	def insert(self, item):
		# insert at end, bubble up is parent is > self
		self.heap.append(item)
		indItem = len(self.heap) - 1
		indParent = self.getParentIndex(indItem)

		while (indParent >= 0 and self.heap[indParent] > self.heap[indItem]):
			swapItem = self.heap[indItem]
			self.heap[indItem] = self.heap[indParent]
			self.heap[indParent] = swapItem

			indItem = indParent
			indParent = self.getParentIndex(indItem)


	def pop(self):
		# get item at top, replace with item at end, bubble down
		returnVal = self.heap[0]
		self.heap[0] = self.heap.pop()

		ind = 0

		while self.getLeftChildIndex(ind) < len(self.heap):
			leftInd = self.getLeftChildIndex(ind)
			rightInd = self.getRightChildIndex(ind)

			if rightInd >= len(self.heap):
				if (self.heap[ind] > self.heap[leftInd]):
					swapItem = self.heap[leftInd]
					self.heap[leftInd] = self.heap[ind]
					self.heap[ind] = swapItem
					ind = leftInd
				else:
					break 

			else:
				if self.heap[rightInd] < self.heap[ind] and self.heap[rightInd] < self.heap[leftInd]:
					swapItem = self.heap[rightInd]
					self.heap[rightInd] = self.heap[ind]
					self.heap[ind] = swapItem
					ind = rightInd
				elif self.heap[leftInd] < self.heap[ind] and self.heap[leftInd] < self.heap[rightInd]:
					swapItem = self.heap[leftInd]
					self.heap[leftInd] = self.heap[ind]
					self.heap[ind] = swapItem
					ind = leftInd 

		return returnVal


	def print(self):
		for i in range(len(self.heap)):
			print(self.heap[i])

		print("End of heap")


def main():
	heap = MinHeap()

	heap.insert(1)
	heap.insert(10)
	heap.insert(11)
	heap.insert(5)
	heap.insert(94)
	heap.insert(-123)
	heap.insert(-2)
	heap.insert(15)

	heap.print()

	print(heap.pop())
	print(heap.pop())
	print(heap.pop())
	print(heap.pop())
	print(heap.pop())
	print("done popping")

	heap.print()


if __name__ == "__main__":
	main()